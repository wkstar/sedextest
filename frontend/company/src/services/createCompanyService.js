import { API_BASE_URL } from '../constants';

const createCompanyService = async (requestData) => {
  const response = await fetch(`${API_BASE_URL}/company`, {
    method: 'POST',
    requestData,
  });

  return await response.json();
};

export default createCompanyService;
