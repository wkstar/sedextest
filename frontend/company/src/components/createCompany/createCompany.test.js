import {
  fireEvent,
  getNodeText,
  render,
  screen,
  waitFor,
} from '@testing-library/react';

import CreateCompany from './CreateCompany';
import createCompanyService from '../../services/createCompanyService';

jest.mock('../../services/createCompanyService');

const getProps = () => ({
  schema: [
    ['companyName', 'Company Name', 'Sedex'],
    ['companyType', 'Company Type', 'NGO'],
  ],
});

describe('createCompany', () => {
  it('renders the form', () => {
    const props = getProps();
    const { asFragment } = render(<CreateCompany {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('validation', () => {
    it('validates missing company type', async () => {
      const props = getProps();
      render(<CreateCompany {...props} />);

      fireEvent.change(screen.getByLabelText('Company Name'), {
        target: { value: 'oo' },
      });

      // Submit form
      fireEvent.click(screen.getByText('Create the Company'));

      await waitFor(() => {
        screen.getByText('Company Type must be set');
      });

      expect(screen.getByText('Company Type must be set')).toBeTruthy();
      expect(
        screen.getByText('Company Name must be greater than 3 characters')
      ).toBeTruthy();
    });
  });

  describe('success', () => {
    it('submits form', async () => {
      createCompanyService.mockResolvedValueOnce({ id: 'ego' });

      const props = getProps();
      render(<CreateCompany {...props} />);

      fireEvent.change(screen.getByLabelText('Company Name'), {
        target: { value: 'BitOfMe Ltd' },
      });

      fireEvent.change(screen.getByLabelText('Company Type'), {
        target: { value: 'SME' },
      });

      // Submit form
      fireEvent.click(screen.getByText('Create the Company'));

      let success;
      await waitFor(() => {
        success = screen.getByTestId('success');
      });

      expect(getNodeText(success)).toEqual(
        'Company with id "ego" successfully created'
      );

      expect(createCompanyService).toHaveBeenCalledTimes(1);
      expect(createCompanyService).toHaveBeenCalledWith({
        companyName: 'BitOfMe Ltd',
        companyType: 'SME',
      });
    });
  });
});
