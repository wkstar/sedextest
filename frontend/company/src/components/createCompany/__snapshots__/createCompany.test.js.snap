// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCompany renders the form 1`] = `
<DocumentFragment>
  <form
    class="css-1l4w6pd"
  >
    <div
      class="css-11818vl"
    >
      <h1
        class="css-0"
      >
        Create a Company
      </h1>
      <div
        class="css-95sj21"
      >
        <div
          class="css-1h3zg0j"
        >
          <label
            for="companyName"
          >
            Company Name
          </label>
        </div>
        <div
          class="css-1h3zg0j"
        >
          <input
            autocomplete="off"
            id="companyName"
            name="companyName"
            placeholder="e.g. Sedex"
          />
        </div>
        <div
          class="css-u41sxy"
        >
          <div
            class="css-1ejw6fi"
          />
        </div>
      </div>
      <div
        class="css-95sj21"
      >
        <div
          class="css-1h3zg0j"
        >
          <label
            for="companyType"
          >
            Company Type
          </label>
        </div>
        <div
          class="css-1h3zg0j"
        >
          <input
            autocomplete="off"
            id="companyType"
            name="companyType"
            placeholder="e.g. NGO"
          />
        </div>
        <div
          class="css-u41sxy"
        >
          <div
            class="css-1ejw6fi"
          />
        </div>
      </div>
      <div
        class="css-95sj21"
      >
        <input
          class="css-y1gt6f"
          type="submit"
          value="Create the Company"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;
